File: tools/r1_scriptbench/w_ui.js
// W3 UI immediate-mode churn — literal translation of w_ui.lua.
// Retained state in a Map keyed by id string (the idiomatic JS equivalent).
"use strict";
const floor = Math.floor;
const NW = 300, FRAMES = 240;

const state = new Map();
let path = "root";

function widget(label, i, f, mx, my) {
  const id = path + "/" + label + i;
  let st = state.get(id);
  if (!st) { st = { scroll: 0, open: false, t: 0 }; state.set(id, st); }
  const x = (i % 3) * 130 + 8;
  const y = floor(i / 3) * 22 + 30 - st.scroll;
  const w = 120, h = 18;
  const hot = mx >= x && mx < x + w && my >= y && my < y + h;
  if (hot) st.t = st.t + 1;
  rect(x, y, w, h, hot ? 0x3a3a52 : 0x2a2a3a);
  rect(x + 2, y + 2, w - 4, h - 4, 0x1a1a24);
  const txt = label + i + ": " + (st.t * 0.1).toFixed(1);
  let tw = 0;
  for (let k = 0; k < txt.length; k++) tw += (txt.charCodeAt(k) < 128) ? 6 : 8;
  st.scroll = (st.scroll + (hot ? 1 : 0)) % 40;
  return tw;
}

const times = new Array(FRAMES);
let acc0 = 0;
for (let f = 1; f <= FRAMES; f++) {
  const t0 = now();
  const mx = (f * 3) % 400;
  const my = (f * 2) % 300;
  let acc = 0;
  for (let i = 0; i < NW; i++) {
    if (i % 50 === 0) path = "root/panel" + floor(i / 50);
    acc += widget("w", i, f, mx, my);
  }
  path = "root";
  times[f - 1] = now() - t0;
  if (f === 1) acc0 = acc;
}

times.sort((a, b) => a - b);
let sum = 0;
for (let i = 0; i < FRAMES; i++) sum += times[i];
emit(`w_ui ${NW} widgets: avg ${(sum / FRAMES / 1000).toFixed(1)} us  p50 ${(times[floor(FRAMES / 2) - 1] / 1000).toFixed(1)}  p99 ${(times[floor(FRAMES * 0.99) - 1] / 1000).toFixed(1)}  max ${(times[FRAMES - 1] / 1000).toFixed(1)}`);

File: tools/r1_scriptbench/w_ui.lua
-- W3 UI immediate-mode churn: 300 widgets/frame x 240 frames.
-- Per widget: id-path string build (push_id shape), per-id retained-state
-- lookup/create, hit-test math, 2 rect() C calls, a formatted value label.
-- Garbage-heavy on purpose: this is where GC pauses show.
local floor = math.floor
local NW, FRAMES = 300, 240

local state = {}   -- per-id retained state, keyed by id path string
local path = "root"

local function widget(label, i, f, mx, my)
  local id = path .. "/" .. label .. i
  local st = state[id]
  if not st then st = { scroll = 0, open = false, t = 0 } state[id] = st end
  local x = (i % 3) * 130 + 8
  local y = floor(i / 3) * 22 + 30 - st.scroll
  local w, h = 120, 18
  local hot = mx >= x and mx < x + w and my >= y and my < y + h
  if hot then st.t = st.t + 1 end
  rect(x, y, w, h, hot and 0x3a3a52 or 0x2a2a3a)
  rect(x + 2, y + 2, w - 4, h - 4, 0x1a1a24)
  local txt = label .. i .. ": " .. string.format("%.1f", st.t * 0.1)
  -- measure stand-in: iterate the label's chars (cm.text measures per glyph)
  local tw = 0
  for k = 1, #txt do tw = tw + ((txt:byte(k) < 128) and 6 or 8) end
  st.scroll = (st.scroll + (hot and 1 or 0)) % 40
  return tw
end

local times = {}
for f = 1, FRAMES do
  local t0 = now()
  local mx = (f * 3) % 400
  local my = (f * 2) % 300
  local acc = 0
  for i = 0, NW - 1 do
    -- panels re-scope the path every 50 widgets (push_id/pop_id shape)
    if i % 50 == 0 then path = "root/panel" .. floor(i / 50) end
    acc = acc + widget("w", i, f, mx, my)
  end
  path = "root"
  times[f] = now() - t0
  if f == 1 then times.acc = acc end
end

table.sort(times)
local sum = 0
for i = 1, FRAMES do sum = sum + times[i] end
emit(string.format("w_ui %d widgets: avg %.1f us  p50 %.1f  p99 %.1f  max %.1f",
  NW, sum / FRAMES / 1000, times[floor(FRAMES / 2)] / 1000,
  times[floor(FRAMES * 0.99)] / 1000, times[FRAMES] / 1000))
